As a user, if I have an evaluation with many features and sampling uncertainty, but one fails due to insufficient data to calculate the optimal block size, I don't want the evaluation to fail

See VLab #120579-257.  The evaluation uses WRES RFC forecasts spanning a year for all NWRFC locations returned by WRDS. The exception occurred for one pool after many features and pools were already complete and led to an overall failure of the evaluation. I was able to reproduce the problem using a two feature run which I'll share in the first comment. The complete exception that led to the evaluation failure is below.

More in comment 1,

Hank

==============================

```
2024-10-18T03:24:35.497+0000  [pool-23-thread-1] INFO EvaluationService - The evaluation failed with the following stack trace: 
wres.pipeline.InternalWresException: Could not complete project execution
    at wres.pipeline.Evaluator.evaluate(Evaluator.java:406)
    at wres.pipeline.Evaluator.evaluate(Evaluator.java:178)
    at wres.Functions.evaluate(Functions.java:135)
    at wres.server.EvaluationService.lambda$startEvaluation$2(EvaluationService.java:695)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: wres.pipeline.WresProcessingException: Encountered an error while processing evaluation 'IYou7J-E__bN5DhqYF7_AZ5Emh4': 
    at wres.pipeline.Evaluator.evaluate(Evaluator.java:835)
    at wres.pipeline.Evaluator.evaluate(Evaluator.java:377)
    ... 7 common frames omitted
Caused by: java.util.concurrent.CompletionException: wres.datamodel.bootstrap.InsufficientDataForResamplingException: Insufficient data to calculate the optimal block size for the stationary bootstrap. The time-series examined had the following metadata: [TimeSeriesMetadata[timeScale=[INSTANTANEOUS],referenceTimes={ISSUED_TIME=2024-06-16T18:23:00Z},variableName=00060,feature=Feature[name=13172500,description=SNAKE RIVER NR MURPHY ID,srid=4326,wkt=POINT ( -116.3906389 43.25481389 )],unit=ft3/s], TimeSeriesMetadata[timeScale=[INSTANTANEOUS],referenceTimes={ISSUED_TIME=2024-06-16T22:22:00Z},variableName=00060,feature=Feature[name=13172500,description=SNAKE RIVER NR MURPHY ID,srid=4326,wkt=POINT ( -116.3906389 43.25481389 )],unit=ft3/s]]
    at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
    at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
    at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1770)
    ... 3 common frames omitted
Caused by: wres.datamodel.bootstrap.InsufficientDataForResamplingException: Insufficient data to calculate the optimal block size for the stationary bootstrap. The time-series examined had the following metadata: [TimeSeriesMetadata[timeScale=[INSTANTANEOUS],referenceTimes={ISSUED_TIME=2024-06-16T18:23:00Z},variableName=00060,feature=Feature[name=13172500,description=SNAKE RIVER NR MURPHY ID,srid=4326,wkt=POINT ( -116.3906389 43.25481389 )],unit=ft3/s], TimeSeriesMetadata[timeScale=[INSTANTANEOUS],referenceTimes={ISSUED_TIME=2024-06-16T22:22:00Z},variableName=00060,feature=Feature[name=13172500,description=SNAKE RIVER NR MURPHY ID,srid=4326,wkt=POINT ( -116.3906389 43.25481389 )],unit=ft3/s]]
    at wres.datamodel.bootstrap.BootstrapUtilities.lambda$getOptimalBlockSizeForStationaryBootstrap$16(BootstrapUtilities.java:459)
    at java.base/java.util.Optional.orElseThrow(Optional.java:403)
    at wres.datamodel.bootstrap.BootstrapUtilities.getOptimalBlockSizeForStationaryBootstrap(BootstrapUtilities.java:459)
    at wres.datamodel.bootstrap.BootstrapUtilities.getOptimalBlockSizeForStationaryBootstrap(BootstrapUtilities.java:62)
    at wres.pipeline.pooling.PoolProcessor.getSamplingUncertaintyStatistics(PoolProcessor.java:555)
    at wres.pipeline.pooling.PoolProcessor.createStatistics(PoolProcessor.java:492)
    at wres.pipeline.pooling.PoolProcessor.get(PoolProcessor.java:151)
    at wres.pipeline.pooling.PoolProcessor.get(PoolProcessor.java:60)
    at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
    ... 3 common frames omitted
```

Reproduced:

```
Caused by: java.util.concurrent.CompletionException: wres.datamodel.bootstrap.InsufficientDataForResamplingException: Insufficient data to calculate the optimal block size for the stationary bootstrap. The time-series examined had the following metadata: [TimeSeriesMetadata[timeScale=[INSTANTANEOUS],referenceTimes={ISSUED_TIME=2024-06-16T22:22:00Z},variableName=00060,feature=Feature[name=13172500,description=SNAKE RIVER NR MURPHY ID,srid=4326,wkt=POINT ( -116.3906389 43.25481389 )],unit=ft3/s], TimeSeriesMetadata[timeScale=[INSTANTANEOUS],referenceTimes={ISSUED_TIME=2024-06-16T18:23:00Z},variableName=00060,feature=Feature[name=13172500,description=SNAKE RIVER NR MURPHY ID,srid=4326,wkt=POINT ( -116.3906389 43.25481389 )],unit=ft3/s]]
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1770)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: wres.datamodel.bootstrap.InsufficientDataForResamplingException: Insufficient data to calculate the optimal block size for the stationary bootstrap. The time-series examined had the following metadata: [TimeSeriesMetadata[timeScale=[INSTANTANEOUS],referenceTimes={ISSUED_TIME=2024-06-16T22:22:00Z},variableName=00060,feature=Feature[name=13172500,description=SNAKE RIVER NR MURPHY ID,srid=4326,wkt=POINT ( -116.3906389 43.25481389 )],unit=ft3/s], TimeSeriesMetadata[timeScale=[INSTANTANEOUS],referenceTimes={ISSUED_TIME=2024-06-16T18:23:00Z},variableName=00060,feature=Feature[name=13172500,description=SNAKE RIVER NR MURPHY ID,srid=4326,wkt=POINT ( -116.3906389 43.25481389 )],unit=ft3/s]]
	at wres.datamodel.bootstrap.BootstrapUtilities.lambda$getOptimalBlockSizeForStationaryBootstrap$16(BootstrapUtilities.java:459)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at wres.datamodel.bootstrap.BootstrapUtilities.getOptimalBlockSizeForStationaryBootstrap(BootstrapUtilities.java:459)
	at wres.datamodel.bootstrap.BootstrapUtilities.getOptimalBlockSizeForStationaryBootstrap(BootstrapUtilities.java:62)
	at wres.pipeline.pooling.PoolProcessor.getSamplingUncertaintyStatistics(PoolProcessor.java:555)
	at wres.pipeline.pooling.PoolProcessor.createStatistics(PoolProcessor.java:492)
	at wres.pipeline.pooling.PoolProcessor.get(PoolProcessor.java:151)
	at wres.pipeline.pooling.PoolProcessor.get(PoolProcessor.java:60)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
```

--- a/wres-datamodel/src/wres/datamodel/bootstrap/BootstrapUtilities.java
+++ b/wres-datamodel/src/wres/datamodel/bootstrap/BootstrapUtilities.java
@@ -47,6 +47,7 @@ public class BootstrapUtilities
      * @param <R> the type of right-ish time-series data
      * @param pool the pool
      * @return the optimal block size for the stationary bootstrap in timestep units, together with the timestep
+     * @throws InsufficientDataForResamplingException if there is insufficient data to calculate the optimal block size
      */
     public static <R> Pair<Long, Duration> getOptimalBlockSizeForStationaryBootstrap( Pool<TimeSeries<Pair<Double, R>>> pool )
     {
@@ -110,7 +111,7 @@ public class BootstrapUtilities
 
     /**
      * Determines whether sufficient data is available for bootstrap resampling. There must be more than one event
-     * across time time-series present.
+     * across the consolidated time-series present.
      * @param <T> the type of time-series data
      * @param data the time-series data
      * @return whether there is sufficient data for the stationary bootstrap
@@ -120,10 +121,13 @@ public class BootstrapUtilities
     {
         Objects.requireNonNull( data );
 
+        // Consolidate the time-series by event datetime and count the number of events
         long eventCount = data.stream()
-                              .mapToLong( t -> t.getEvents()
-                                                .size() )
-                              .sum();
+                              .flatMap( t -> t.getEvents()
+                                              .stream()
+                                              .map( Event::getTime ) )
+                              .distinct()
+                              .count();
 
         if ( LOGGER.isDebugEnabled() )
         {
@@ -413,6 +417,7 @@ public class BootstrapUtilities
      * @param <T> the type of time-series data
      * @param pool the pool
      * @return the optimal block size for the stationary bootstrap
+     * @throws InsufficientDataForResamplingException if there are fewer than two events across the consolidated series
      */
     private static <T> Pair<Long, Duration> getOptimalBlockSizeForStationaryBootstrap( List<TimeSeries<Pair<Double, T>>> pool )
     {

--- a/wres-datamodel/test/wres/datamodel/time/TimeSeriesTest.java
+++ b/wres-datamodel/test/wres/datamodel/time/TimeSeriesTest.java
@@ -26,7 +26,7 @@ import wres.statistics.generated.ReferenceTime.ReferenceTimeType;
 
 /**
  * Tests the {@link TimeSeries}
- * 
+ *
  * @author James Brown
  */
 
@@ -79,10 +79,10 @@ public class TimeSeriesTest
                                        FEATURE_NAME,
                                        UNIT );
         this.testSeries = builder.setMetadata( metadata )
-                            .addEvent( iterator.next() )
-                            .addEvent( iterator.next() )
-                            .addEvent( iterator.next() )
-                            .build();
+                                 .addEvent( iterator.next() )
+                                 .addEvent( iterator.next() )
+                                 .addEvent( iterator.next() )
+                                 .build();
     }
 
     /**
@@ -146,7 +146,7 @@ public class TimeSeriesTest
     @Test
     public void testEquals()
     {
-        // Reflexive 
+        // Reflexive
         assertEquals( this.testSeries, this.testSeries );
 
         // Symmetric

